### PR TITLE
Reduce aligned series compaction memory usage

### PR DESCRIPTION
**Pull request by shuwenwei**
## Description
Reduce aligned series compaction memory usage
Relevant modifications in tsfile repository: https://github.com/apache/tsfile/pull/151
截屏2024-07-19 16 52 53

### CODE REVIEW

File: iotdb-core/node-commons/src/assembly/resources/conf/iotdb-system.properties.template [hunk 1]
```diff
@@ -1270,6 +1270,12 @@ min_cross_compaction_unseq_file_level=1
 # Datatype: int
 compaction_thread_count=10
 
+# How many chunk will be compact in aligned series compaction, 10 by default.
+# Set to Integer.MAX_VALUE when less than or equal to 0.
+# effectiveMode: hot_reload
+# Datatype: int
+compaction_max_aligned_series_num_in_one_batch=10
```
Comment (jt2594838) [inline iotdb-core/node-commons/src/assembly/resources/conf/iotdb-system.properties.template:1277]: chunk -> chunks
compact -> compacted

**Review by jt2594838 — APPROVED** [review]

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/selector/estimator/FastCompactionInnerCompactionEstimator.java [hunk 2]
```diff
@@ -76,4 +83,18 @@ public long calculatingDataMemoryCost(CompactionTaskInfo taskInfo) throws IOExce
...
+      throws IOException {
+    int maxConcurrentSeriesNum =
+        Math.max(
+            config.getCompactionMaxAlignedSeriesNumInOneBatch(), config.getSubCompactionTaskNum());
+    long maxChunkSize = config.getTargetChunkSize();
+    long maxPageSize = tsFileConfig.getPageSizeInByte();
+    int maxOverlapFileNum = calculatingMaxOverlapFileNumInSubCompactionTask(resources);
+    // source files (chunk + uncompressed page) * overlap file num
+    // target file (chunk + unsealed page writer)
+    return (maxOverlapFileNum + 1) * maxConcurrentSeriesNum * (maxChunkSize + maxPageSize);
```
[the remark this answers is not in the document]
Comment (shuwenwei) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/selector/estimator/FastCompactionInnerCompactionEstimator.java:98]: Unsequence inner space compaction have overlapping files

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/load/TsFileSplitter.java [hunk 3]
```diff
@@ -318,6 +145,234 @@ public void splitTsFileByDataPartition() throws IOException, IllegalStateExcepti
...
+            isAligned, ((PlainDeviceID) curDevice).toStringID(), header, timePartitionSlot);
+
+    if (!needDecodeChunk(chunkMetadata)) {
+      chunkData.setNotDecode();
+      chunkData.writeEntireChunk(reader.readChunk(-1, header.getDataSize()), chunkMetadata);
+      if (isAligned) {
+        isTimeChunkNeedDecode = false;
+        pageIndex2ChunkData
+            .computeIfAbsent(1, o -> new ArrayList<>())
+            .add((AlignedChunkData) chunkData);
```
[the remark this answers is not in the document]
Comment (shuwenwei) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/execution/load/TsFileSplitter.java:187]: The pageIndex2ChunkData is a map, and it use '1' as first element.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/executor/batch/BatchedFastAlignedSeriesCompactionExecutor.java [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+      throws IOException {
+    // 1. get Full AlignedChunkMetadata from cache
+    List<AlignedChunkMetadata> alignedChunkMetadataList = null;
+    if (alignedChunkMetadataCache.containsKey(tsFileResource)) {
+      alignedChunkMetadataList = alignedChunkMetadataCache.get(tsFileResource);
+    } else {
+      alignedChunkMetadataList = getAlignedChunkMetadataList(tsFileResource);
+      AlignedSeriesBatchCompactionUtils.markAlignedChunkHasDeletion(alignedChunkMetadataList);
+      alignedChunkMetadataCache.put(tsFileResource, alignedChunkMetadataList);
+    }
```
[the remark this answers is not in the document]
Comment (shuwenwei) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/executor/batch/BatchedFastAlignedSeriesCompactionExecutor.java:120]: The method 'getAlignedChunkMetadataList' will throw an IOException.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/executor/batch/utils/AlignedSeriesBatchCompactionUtils.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+      compactedMeasurements.add(schema.getMeasurementId());
+      if (selectedColumnBatch.size() >= batchSize) {
+        break;
+      }
+      if (compactedMeasurements.size() == schemaList.size()) {
+        break;
+      }
+    }
+    return selectedColumnBatch;
+  }
```
[the remark this answers is not in the document]
Comment (shuwenwei) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/executor/batch/utils/AlignedSeriesBatchCompactionUtils.java:78]: We can not get chunk size before read ChunkHeader. It is hard to get chunk size of each series.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/executor/batch/utils/AlignedSeriesBatchCompactionUtils.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+          return ModifiedStatus.ALL_DELETED;
+        }
+        if (range.overlaps(new TimeRange(startTime, endTime))) {
+          // exist data on this page or chunk been deleted
+          status = ModifiedStatus.PARTIAL_DELETED;
+        }
+      }
+    }
+    return status;
+  }
```
[the remark this answers is not in the document]
Comment (shuwenwei) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/executor/batch/utils/AlignedSeriesBatchCompactionUtils.java:174]: Maybe it can merge consecutive Deletions when setting the deletion interval list for ChunkMetadata?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/executor/batch/utils/AlignedSeriesBatchCompactionUtils.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    for (IChunkMetadata chunkMetadata : alignedChunkMetadata.getValueChunkMetadataList()) {
+      if (chunkMetadata == null) {
+        continue;
+      }
+      if (measurementIndex.containsKey(chunkMetadata.getMeasurementUid())) {
+        valueChunkMetadataList.set(
+            measurementIndex.get(chunkMetadata.getMeasurementUid()), chunkMetadata);
+      }
```
[the remark this answers is not in the document]
Comment (shuwenwei) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/executor/batch/utils/AlignedSeriesBatchCompactionUtils.java:125]: The measurementIndex is not modified here, but the valueChunkMetadataList is modified. Using computeIfPresent may seem a bit strange.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/executor/batch/utils/AlignedSeriesBatchCompactionUtils.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jt2594838) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/executor/batch/utils/AlignedSeriesBatchCompactionUtils.java:78]: Maybe we can record the maximum chunk size of each device or of the a file in TsFileResource?
It may be helpful to perform more strict memory control.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/executor/batch/utils/AlignedSeriesBatchCompactionUtils.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (jt2594838) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/executor/batch/utils/AlignedSeriesBatchCompactionUtils.java:125]: Not modifying measurementIndex , just to avoid accessing the map twice.
```suggestion
      chunkMetadataMap.computeIfPresent(chunkMetadata.getMeasurementUid(),
        (k, v) -> {
          valueChunkMetadataList.add(v);
          return v;
        }
    );
```

[the remark this answers is not in the document]
Comment (jt2594838) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/executor/batch/utils/AlignedSeriesBatchCompactionUtils.java:125]: Not modifying measurementIndex , just to avoid accessing the map twice.
```suggestion
      chunkMetadataMap.computeIfPresent(chunkMetadata.getMeasurementUid(),
        (k, v) -> {
          valueChunkMetadataList.add(v);
          return v;
        }
    );
```

**Review by jt2594838 — APPROVED** [review]

**Review by jt2594838 — APPROVED** [review]

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/writer/AbstractCompactionWriter.java [hunk 8]
```diff
@@ -41,6 +41,7 @@
 import org.apache.tsfile.write.chunk.ChunkWriterImpl;
 import org.apache.tsfile.write.chunk.IChunkWriter;
 import org.apache.tsfile.write.chunk.ValueChunkWriter;
+import org.apache.tsfile.write.schema.Schema;
```
Comment (jt2594838) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/compaction/execute/utils/writer/AbstractCompactionWriter.java:44]: Why is this import added but not used?
